### PR TITLE
Fix several small issues with shared caching

### PR DESCRIPTION
**Pull request by JakeSiFive**
These were all known issues that we haven't hit testing against real code yet but which I believe will come up. At this time this represents all known bugs with shared caching. The next step is to test it more widely on more code and to fuzz test the cache type in C++.

Issues fixed:
* Output directories would be ignored so if a tool created a directory with a non-standard mode or if it that created a directory that contained no files that directory would not be created. We now store these in the cache database and recreate them faithfully. If we did have to hash them they would hash to all zeros but we don't need to do that.
* Symlinks would have been hashed to the content of the file they pointed to and the content of the file they pointed to would have been copied instead. We store the readlink value in the cache database and recreate these faithfully now. We don't have to worry about hashing these because on input they're handled like anything else and on output shim-wake will later hash them instead of us.
* We had some code to speed up hashing of files by doing them in parallel but that introduced new threads which would mess with signal handling. I undid this. We took a big performance hit (on one real example we went from taking 22 seconds to 45 seconds). This is unfortunate but not a deal breaker since its still saving time over what would be a 10 minute job otherwise. We can optimize things later.

### CODE REVIEW

File: src/job-cache/job-cache.cpp
Comment (mmjconolly) [inline src/job-cache/job-cache.cpp:89]: I'd have assumed that the shared-cache were linux-only, as the macos implementation of wakebox can't hide the .fuse directory when a payload command resolves an absolute path

Comment (JakeSiFive) [reply src/job-cache/job-cache.cpp:89]: The shared cache itself should be able to work on mac but that may be an issue that I was about to spend a lot time debugging. The tests failed when they ran on mac despite succeeding on the tarball step. I guess I need to update these tests to pass without testing on mac.

Comment (JakeSiFive) [reply src/job-cache/job-cache.cpp:89]: Thanks for the heads up!

**Comment by JakeSiFive** [comment]
BTW, on the slow down. The machine I was testing on was actually loaded. Reads should be just as fast. I was thinking about it but the hashing only happens on write not on read so there should be no slow down on read and I'm not sure we care too much about the slow down on write since the write is non-blocking.

File: tests/job-cache/basic-fetch/pass.sh
Comment (mmjconolly) [inline tests/job-cache/basic-fetch/pass.sh:3]:
```suggestion
if [ $(uname) != Linux ] ; then
```

Comment (JakeSiFive) [reply tests/job-cache/basic-fetch/pass.sh]: Good call, thanks!

File: src/job-cache/job-cache.h
Comment (V-FEXrt) [inline src/job-cache/job-cache.h:141]: this is done no?

File: src/job-cache/job-cache.cpp
Comment (V-FEXrt) [inline src/job-cache/job-cache.cpp:171]: this looks eerily similar to writing a raw NAND device driver lol

File: src/job-cache/job-cache.cpp
Comment (V-FEXrt) [inline src/job-cache/job-cache.cpp:906]: typo: `hasing` -> `hashing`

File: src/job-cache/job-cache.cpp
Comment (V-FEXrt) [inline src/job-cache/job-cache.cpp:941]: Do we expect a bunch of things that aren't dirs, links, or regular files? 

if not, might be good to add a debug warning before the continue

File: src/job-cache/job-cache.h
Comment (JakeSiFive) [inline src/job-cache/job-cache.h:141]: Ah crap I did this one already, I'll remove that TODO before this lands.

File: src/job-cache/job-cache.cpp
Comment (V-FEXrt) [inline src/job-cache/job-cache.cpp:1109]: typo: `witht` -> `with`

File: src/job-cache/job-cache.cpp
Comment (V-FEXrt) [inline src/job-cache/job-cache.cpp:1112]: typo: `avliable` -> `available`

File: src/job-cache/job-cache.cpp
Comment (V-FEXrt) [inline src/job-cache/job-cache.cpp:1125]: typo: `avliable` -> `available`

File: src/job-cache/job-cache.cpp
Comment (V-FEXrt) [inline src/job-cache/job-cache.cpp:1189]: what is this code doing?

Comment (JakeSiFive) [reply src/job-cache/job-cache.cpp:1189]: This is one of those times you're baiting me to add a comment huh ;)

These rename sandbox paths into non-sandbox paths. The paths that I expect in the json are sandbox absolute paths and thats also what I store in the database. The paths outside the sandbox are generally relative. You can add redirects to redirect any sandbox path to a non-sandbox location. For instance you can redirect "/workspace" to "./"

File: src/job-cache/job-cache.cpp
Comment (JakeSiFive) [reply src/job-cache/job-cache.cpp:941]: No I don't think so. The only other options are things like devices, character devices (I think that's a terminal?), block devices, and sockets...stuff that you can't cache even if you wanted to and stuff that fuse isn't going to tell us about.

File: src/job-cache/job-cache.cpp
Comment (JakeSiFive) [reply src/job-cache/job-cache.cpp:171]: heh. Hopefully we don't have the issue where the first 4 physical blocks were being used for critical stuff and they were getting overloaded because the designers were like "it will probably never have that many cycles on it" and then all of our infra started to fail one by one at the exact same time...yeah that wasn't fun.

File: src/job-cache/job-cache.h
Comment (JakeSiFive) [reply src/job-cache/job-cache.h:141]: yeah I'll remove it

File: src/job-cache/job-cache.h
Comment (JakeSiFive) [reply src/job-cache/job-cache.h:128]: OutputSymlink has the path to the symlink and the value you'd get from calling readlink on that symlink. OutputDirectory just has the path and the permission bits for the directory.

File: .gitignore
Comment (JakeSiFive) [reply .gitignore:46]: That's probably a better idea huh?

File: src/job-cache/job-cache.cpp
Comment (JakeSiFive) [reply src/job-cache/job-cache.cpp]: I added comments, let me know if that helps.

Comment (V-FEXrt) [reply src/job-cache/job-cache.cpp]: eh goal wasn't directly to bait you into writing a comment but I did have trouble following what it was doing so a comment or function extraction felt reasonable

File: src/job-cache/job-cache.cpp
Comment (JakeSiFive) [reply src/job-cache/job-cache.cpp:940]: I can and probably should extract this out into a function.

File: src/job-cache/job-cache.cpp
Comment (V-FEXrt) [reply src/job-cache/job-cache.cpp:171]: 👀

File: src/job-cache/job-cache.cpp
Comment (V-FEXrt) [reply src/job-cache/job-cache.cpp:940]: I'll not block on it, but I'll strongly urge you to consider it :)

**Review by V-FEXrt — APPROVED** [review]

File: src/job-cache/job-cache.cpp
Comment (JakeSiFive) [reply src/job-cache/job-cache.cpp:940]: Yeah I already did it and its upload now :) I just had stuff to do this morning so it took me a while to finish.

**Pull request merged**
